comms/src/bridge/Connection.ts

```diff
@@ -0,0 +1,110 @@
+import {ChildProcess} from 'child_process';
+import {Duplex} from 'stream';
+import {SerialPort} from 'serialport';
+import {MessageQueue} from '../parse/MessageQueue';
+import {Request} from '../parse/Requests';
+
+type PromiseResolver<R> = (value: R | PromiseLike<R>) => void;
+
+export abstract class Connection {
+    protected channel: Duplex;
+
+    protected requests: [Request<any>, PromiseResolver<any>][];
+
+    protected messages: MessageQueue;
+
+    protected constructor(channel: Duplex) {
+        this.channel = channel;
+        this.requests = [];
+        this.messages = new MessageQueue('\n');
+
+        this.listen();
+    }
+
+    // listen on duplex channel
+    protected listen(): void {
+        this.channel.on('data', (data: Buffer) => {
+            console.log(data.toString());
+            this.messages.push(data.toString());
+            this.process();
+        });
+    }
+
+    // process messages in queue
+    protected process(): void {
+        // until no complete messages are left
+        for (let message of this.messages) {
+            const index: number = this.search(message);  // search request
+
+            if (0 <= index && index < this.requests.length) {
+                // parse and resolve
+                const [candidate, resolver] = this.requests[index];
+                resolver(candidate.parser(message));
+                this.requests.splice(index, 1);  // delete resolved request
+            }
+
+            // silently drop message if no requests match
```

Comment:
I think it should be possible to do both: silently drop messages and register a catch-all handler. The latter is particularly important when debugging.

Reply:
> The latter is particularly important when debugging.
I don't immediately why. But I think this can be added on the other branch if necessary.